crates/chia-consensus/src/spendbundle_validation.rs

```diff
@@ -0,0 +1,416 @@
+use crate::gen::flags::{ALLOW_BACKREFS, DISALLOW_INFINITY_G1, ENABLE_MESSAGE_CONDITIONS};
+use crate::gen::opcodes::{
+    AGG_SIG_AMOUNT, AGG_SIG_ME, AGG_SIG_PARENT, AGG_SIG_PARENT_AMOUNT, AGG_SIG_PARENT_PUZZLE,
+    AGG_SIG_PUZZLE, AGG_SIG_PUZZLE_AMOUNT,
+};
+use crate::gen::owned_conditions::OwnedSpendBundleConditions;
+use crate::gen::validation_error::ErrorCode;
+use crate::spendbundle_conditions::get_conditions_from_spendbundle;
+use chia_bls::BlsCache;
+use chia_bls::PairingInfo;
+use chia_protocol::SpendBundle;
+use clvmr::{ENABLE_BLS_OPS_OUTSIDE_GUARD, ENABLE_FIXED_DIV, LIMIT_HEAP};
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+
+// currently in mempool_manager.py
+// called in threads from pre_validate_spend_bundle()
+// returns (error, cached_results, new_cache_entries, duration)
+pub fn validate_clvm_and_signature(
+    spend_bundle: &SpendBundle,
+    max_cost: u64,
+    constants: &ConsensusConstants,
+    height: u32,
+    cache: &Arc<Mutex<BlsCache>>,
+) -> Result<(OwnedSpendBundleConditions, Vec<PairingInfo>, Duration), ErrorCode> {
+    let start_time = Instant::now();
+    let mut a = make_allocator(LIMIT_HEAP);
+    let sbc = get_conditions_from_spendbundle(&mut a, spend_bundle, max_cost, height, constants)
+        .map_err(|e| e.1)?;
+    let npcresult = OwnedSpendBundleConditions::from(&a, sbc);
+    let iter = npcresult.spends.iter().flat_map(|spend| {
+        let condition_items_pairs = [
+            (AGG_SIG_PARENT, &spend.agg_sig_parent),
+            (AGG_SIG_PUZZLE, &spend.agg_sig_puzzle),
+            (AGG_SIG_AMOUNT, &spend.agg_sig_amount),
+            (AGG_SIG_PUZZLE_AMOUNT, &spend.agg_sig_puzzle_amount),
+            (AGG_SIG_PARENT_AMOUNT, &spend.agg_sig_parent_amount),
+            (AGG_SIG_PARENT_PUZZLE, &spend.agg_sig_parent_puzzle),
+            (AGG_SIG_ME, &spend.agg_sig_me),
+        ];
+        condition_items_pairs
+            .into_iter()
+            .flat_map(move |(condition, items)| {
+                let spend_clone = spend.clone();
+                items.iter().map(move |(pk, msg)| {
+                    (
+                        pk,
+                        make_aggsig_final_message(
+                            condition,
+                            msg.as_slice(),
+                            &spend_clone,
+                            constants,
+                        ),
+                    )
+                })
+            })
+    });
+    let unsafe_items = npcresult
+        .agg_sig_unsafe
+        .iter()
+        .map(|(pk, msg)| (pk, msg.as_slice().to_vec()));
+    let iter = iter.chain(unsafe_items);
+    // Verify aggregated signature
+    let (result, added) = cache
+        .lock()
+        .unwrap()
+        .aggregate_verify(iter, &spend_bundle.aggregated_signature);
+    if !result {
+        return Err(ErrorCode::BadAggregateSignature);
+    }
+    Ok((npcresult, added, start_time.elapsed()))
```

Comment:
```suggestion
    Ok((npcresult, cache.items(), start_time.elapsed()))
```

This change requires you to add the `items()` function to the cache (which currently only exists in the python binding). Alternatively (but maybe a bigger change) would be to replace the call to `aggregate_verify()` with just a loop over computing the pairings, and return those. There's already an `aggregate_verify_gt()` for the signature validation